File: cmps101/pa5/List.c
/*
Tim Mertogul
1331402
tmertogu
CMPS101 Summer 2014
08/11/14
PA5
List.c
List ADT
*/

#include<stdio.h>
#include<stdlib.h>
#include "List.h"


// structs --------------------------------------------------------------------

// private NodeObj type
typedef struct NodeObj{
   int data;
	//int index;
   struct NodeObj* next;
   struct NodeObj* prev;
} NodeObj;

// private Node type
typedef NodeObj* Node;

// private ListObj type
typedef struct ListObj{
   Node front;
   Node back;
   Node cursor;
   int length;
   int index;
} ListObj;


// Constructors-Destructors ---------------------------------------------------

// newNode()
// Returns reference to new Node object. Initializes next and data fields.
// Private.
Node newNode(int data){
   Node N = malloc(sizeof(Node));
   N->data = data;
   N->next = NULL;
   N->prev = NULL;
   return(N);
}

// freeNode()
// Frees heap memory pointed to by *pN, sets *pN to NULL.
// Private.
void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}

// newList()
// Returns reference to new empty List object.
List newList(void){
   List L = malloc(sizeof(ListObj));
   L->front = L->back = NULL;
   L->length = 0;
   L->cursor = NULL;
   L->index = -1;
   return(L);
}


// freeList()
// Frees all heap memory associated with List *pL, and sets *pL to NULL.S
void freeList(List* pL){
	if(pL!=NULL || *pL!=NULL) { 
		return;
	}
	while( !isEmpty(*pL) ) { 
		deleteFront(*pL); 
	}
	free(*pL);
	*pL = NULL;
}

// Access functions -----------------------------------------------------------

// getLength()
// Returns the length of L.
int length(List L){
   return(L->length);
}

// getIndex()
// Returns the index of L.
int getIndex(List L){
    if( off(L) ) {
        L->index = -1;
    }
    if( isEmpty(L) ){ 
        L->index = -1;
    }
   return(L->index);
}

//returns front element in this list Pre: lengh()>0
int front(List L){
	if( L == NULL ){
		printf("List Error: called front() on null List\n");
		exit(1);	
    }
   if( isEmpty(L) ){
	  printf("List Error: calling front() on an empty List\n");
	  exit(1);
   }
   return(L->front->data);

}

//returns back element in this list Pre: lengh()>0
int back(List L){
	if( L == NULL ){
		printf("List Error: called back() on null List\n");
		exit(1);	
    }
   if( isEmpty(L) ){
	  printf("List Error: calling back() on an empty List\n");
	  exit(1);
   }
   return(L->back->data);
}

//returns cursor element in this list Pre: lengh()>0, getIndex()>=0
int getElement(List L){
	if( L == NULL ){
		printf("List Error: called getElement() on null List\n");
		exit(1);	
    }
   if( isEmpty(L) ){
      printf("List Error: calling getElement() on an empty List\n");
      exit(1);
   }
   return(L->cursor->data);
}


// isEmpty()
int isEmpty(List L){
    if( L == NULL ){
        exit(1);
    }
    return(L->length==0);
}

//Returns true if this List and L are the same integer sequence. The cursor is ignored in both lists
int equals(List A, List B){
   int flag = 1;
   Node N = NULL;
   Node M = NULL;

   if( A==NULL || B==NULL ){
      printf("List Error: calling equals() on NULL List\n");
      exit(1);
   } 
   N = A->front;
   M = B->front;
   if( A->length != B->length ) { return 0; }
   while( flag && N!=NULL){
      flag = (N->data==M->data);
      N = N->next;
      M = M->next;
   }
   return flag;
}


// Manipulation procedures ----------------------------------------------------

// Re-sets this List to the empty state.
void clear(List L){
	Node temp;
	while( L->front != NULL ){
		temp = L->front;
		L->front = L->front->prev;
		freeNode(&temp);
	}
	L->index = -1;
	L->length = 0;
}

// If 0<=i<=length()-1, moves the cursor to the element at index i, otherwise the cursor becomes undefined.
void moveTo(List L, int i) {
	
	if (0 <= i && i <= L->length -1){
		L->cursor = L->front;
		L->index = 0;
		for(int j = 0; j < i; j++){
			L->cursor = L->cursor->prev;
			L->index++;
		}

	} else { 
		//System.out.println("nulling");
		L->cursor = NULL; 
		L->index = -1;
	}	
}

// If 0<getIndex()<=length()-1, moves the cursor one step toward the front of the list. If getIndex()==0, cursor becomes undefined.  If getIndex()==-1, cursor remains undefined. 
void movePrev(List L){

	if(L->index > L->length -1){
		L->cursor = NULL;
		L->index = -1;
	} else if (L->index == -1 || L->index == 0){
		L->cursor = NULL;
		L->index = -1;
	} else { 
		L->cursor = L->cursor->prev;
		L->index--;
	}
}

// If 0<=getIndex()<length()-1, moves the cursor one step toward the back of the list. If getIndex()==length()-1, cursor becomes undefined. If index==-1, cursor remains undefined.
void moveNext(List L){
	if(L->index == L->length-1){
		L->cursor = NULL;
		L->index = -1;
	} else if (L->index == -1){
		L->cursor = NULL;
	} else {
		L->cursor = L->cursor->next;
		L->index++;
	}
}

// Inserts new element before front element in this List.
void prepend(List L, int data) {
	Node newF = newNode(data);
	if(isEmpty(L)){ 
		L->front = L->back = newF;
	}
	else { 
		newF->next = L->front;
		L->front->prev = newF;
		L->front = newF;
	}
	L->length++;	
	if (getIndex(L) != -1) {
		L->index++;
	}
	//freeNode(&newF);
}

// Inserts new element after back element in this List.
void append(List L, int data) { 
	Node newB = newNode(data);
	if(isEmpty(L)){ 
		//printf("empty, adding %d\n", data);
		L->front = L->back = newB;
	}
	else{
		//printf("appending %d\n", data);
		newB->prev = L->back;
		L->back->next = newB;
		L->back = newB;
	}
	L->length++;
	//freeNode(&newB);
	//printf("appended %d\n", data);
}


// Inserts new element before cursor element in this List. Pre: length()>0, getIndex()>=0
void insertBefore(List L, int data){
	if(getIndex(L) == 0){
		prepend(L,data);
	}else if(getIndex(L) >= 0 && length(L) > 0 ){
		Node newBC = newNode(data);
		Node temp = NULL;
		temp = L->cursor->prev;
		newBC->prev = temp;
		temp->next = newBC;
		newBC->next = L->cursor;
		L->cursor->prev = newBC;
		L->length++;
		if (getIndex(L) != -1) {
			L->index++;
		}	
		freeNode(&newBC);	
	}
	else{
	}
}

// Inserts new element after cursor element in this List. Pre: length()>0, getIndex()>=0
void insertAfter(List L, int data){
	if(getIndex(L) == length(L)-1 ){
		append(L,data);
	}
	else{
		Node newAC = newNode(data);
		Node temp = NULL;
		temp = L->cursor->next;
		newAC->next = temp;
		temp->prev = newAC;
		newAC->prev = L->cursor;
		L->cursor->next = newAC;
		L->length++;	
		freeNode(&newAC);
	}		
}


// Deletes the front element in this List. Pre: length()>0
void deleteFront(List L){
	Node temp = NULL;
	temp = L->front->next;
	L->front = temp;	
	L->front->prev = NULL;	
	temp = NULL;
	L->length--;
	
	if (getIndex(L) != -1) {
		L->index--;
	}
	freeNode(&temp);
}

// Deletes the back element in this List. Pre: length()>0
void deleteBack(List L){  
	Node temp = NULL;
	temp = L->back->prev;
	L->back = temp;
	L->back->next = NULL;
	temp = NULL;
	L->length--;
	freeNode(&temp);
}

//Deletes cursor element in this List. Cursor is undefined. Pre: length()>0, getIndex()>=0
void delete(List L){
	Node be4;
	Node l8r;
	//System.out.println("length ="+length());
	if(getIndex(L) == 0){
		deleteFront(L);
		L->length++;
		//front = front.next;
	} else if(getIndex(L) == length(L)-1){
		//System.out.println("deleting back =");
		deleteBack(L);
		L->length++;
		//back = back.prev;
	} else {
		be4 = L->cursor->prev;
		l8r = L->cursor->next;
		be4->next = l8r;	
		l8r->prev = be4;
	}
	L->cursor = NULL;
	L->length--;
	freeNode(&be4);
	freeNode(&l8r);
} 

int off(List L){
	if( L->cursor == NULL ){
		return(1);
	}
	else return(0);
} 

// Other Functions ------------------------------------------------------------

// printList()
// Prints data elements in L on a single line to stdout.
void printList(FILE* out, List L){
   if( L==NULL ){
      fprintf(out, "List Error: calling printList() on NULL List reference\n");
      exit(1);
   }
   NodeObj* N = NULL;
   for(N = L->front; N != NULL; N = N->next){
      fprintf(out, "%d ", N->data);
   }
}


List copyList(List L){
	List sb = newList();
	//printf("moving to 0 \n");
	moveTo(L,0);
	//printf("MOVED to 0 \n");
	while(getIndex(L) < length(L)-1){
		append(sb,getElement(L));
		moveNext(L);
	}
	if(length(sb) != 0){
		append(sb,getElement(L));
	}
	L->cursor = NULL;
	sb->cursor = NULL;
	return sb;
	
}

File: cmps101/pa3/List.java
/*Tim Mertogul
* 1331402
* CMPS101 Summer14
* 07/02/14
* List ADT for pa3
* List.java
*/

public class List {
	
	private class Node {
		Object MatrixEntry;        //not sure what for
		//int index; 
		Node prev;
		Node next;
		Node cursor;  //current value
		
		Node(Object MatrixEntry){  //constructor
			//this.element = element;
			this.MatrixEntry = MatrixEntry;
			//this.index = index;
			this.next = next;
			this.prev = prev;
			//this.index = index;
		}
		
	}
	private Node front;     // reference to first Node in List
	private Node back;     // reference to last Node in List
 	private Node cursor;     // reference to current Node in List
 	private int length;  // length of list
	private int index;
	
	//constructor
	public List(){
	    front = back = cursor = null;
		index = -1;
	    length = 0;
	}
	
	//returns number of elements in this List
	int length() {
		return length;
	}
	
	//returns the index of the cursor element in this list or returns -1 if the cursor is undefined
	int getIndex(){
		if (cursor == null) {
			return -1;
		}
		else {
			return index;
		}
	}	
	
	//returns front element in this list Pre: lengh()>0
	Object front(){ 
		if(length() <= 0){
			throw new RuntimeException("List Error: front() called on empty List");
		} else {
			return front.MatrixEntry; 
		}
	}
	
	//returns back element in this list Pre: lengh()>0
	Object back(){ 
		if(length() <= 0){
			throw new RuntimeException("List Error: back() called on empty List");
		} else {
			return back.MatrixEntry; 
		}
	}
	
	//returns cursor element in this list Pre: lengh()>0, getIndex()>=0
	Object getElement(){
		if(length() <= 0 || getIndex() < 0){
			throw new RuntimeException("List Error: getElement() called on empty List or cursor points to null");
		} else {
			return cursor.MatrixEntry; 
		}
	}
	
	//Returns true if this List and L are the same integer sequence. The cursor is ignored in both lists
	public boolean equals(Object x){
		
		/*
		if(length() != L.length()){
			return false;
		}
		moveTo(0);
		L.moveTo(0);
		for(getIndex(); getIndex() < length()-1; moveNext()){
			if( getElement() != L.getElement()){
				return false;
			}
			L.moveNext();
			if( back() != L.back()){
				return false;
			}
		}*/
		return true;
	}
	
	// Re-sets this List to the empty state.
	void clear(){ 
		front = back = cursor = null;
		index = -1;
		length = 0;
	}
	
	// If 0<=i<=length()-1, moves the cursor to the element at index i, otherwise the cursor becomes undefined.
	void moveTo(int i) {
		Node temp = null;
		if (0 <= i && i <= length()-1){
			cursor = front;
			index = 0;
			while(getIndex() < i){
				moveNext();
			}

		} else { 
			cursor = null; 
			index = -1;
		}
	}
	
	// If 0<getIndex()<=length()-1, moves the cursor one step toward the front of the list. If getIndex()==0, cursor becomes undefined.  If getIndex()==-1, cursor remains undefined. 
	void movePrev(){
		if(getIndex() > length()-1){
			cursor = null;
			index = -1;
		} else if (getIndex() == -1 || getIndex() == 0){
			cursor = null;
			index = -1;
		} else { 
			cursor = cursor.prev;
			index--;
		}
	}
			
	// If 0<=getIndex()<length()-1, moves the cursor one step toward the back of the list. If getIndex()==length()-1, cursor becomes undefined. If index==-1, cursor remains undefined.
	void moveNext() {
		if(getIndex() == length()-1){
			cursor = null;
			index = -1;
		} else if (getIndex() == -1){
			cursor = null;
			index = -1;
		} else {
			cursor = cursor.next;
			index++;
		}
	}
	
	// Inserts new element before front element in this List.
	void prepend(Object MatrixEntry) {
		Node newF = new Node(MatrixEntry);
		//Node temp = null;
		if(length() ==0){ 
			back = newF;
		}
		else { 
			front.prev = newF;
			newF.next = front;
		}
		front = newF;
		length++;	
		if (getIndex() != -1) {
			index++;
		}
	}
	
	// Inserts new element after back element in this List.
	void append(Object MatrixEntry) { 
		Node newB = new Node(MatrixEntry);
		if(length() ==0){ 
			front = newB;
		}
		else{
		    back.next = newB;
		    newB.prev = back;
		}
		back = newB;
		length++;	
	}	
	
	// Inserts new element before cursor element in this List. Pre: length()>0, getIndex()>=0
	
	void insertBefore(Object MatrixEntry){
		if (length()<=0 || getIndex()<0) {
			throw new RuntimeException("List Error: insertBefore() called on empty List or cursor points to null");
		}
		if(getIndex() == 0){
			prepend(MatrixEntry);
		}else if(getIndex() >= 0 && length() > 0 ){
			Node newBC = new Node(MatrixEntry);
			Node temp = null;
			temp = cursor.prev;
			newBC.prev = temp;
			temp.next = newBC;
			newBC.next = cursor;
			cursor.prev = newBC;
			length++;
			if (getIndex() != -1) {
				index++;
			}		
		}	
	}
	
	// Inserts new element after cursor element in this List. Pre: length()>0, getIndex()>=0
	void insertAfter(Object MatrixEntry){
		if (length()<=0 || getIndex()<0) {
			throw new RuntimeException("List Error: insertAfter() called on empty List or cursor points to null");
		}
		if(getIndex() == length()-1 ){
			append(MatrixEntry);
		}
		else{
			Node newAC = new Node(MatrixEntry);
			Node temp = null;
			temp = cursor.next;
			newAC.next = temp;
			temp.prev = newAC;
			newAC.prev = cursor;
			cursor.next = newAC;
			length++;	
		}
	}
	
	// Deletes the front element in this List. Pre: length()>0
	void deleteFront(){
		if (length()<=0 || getIndex()<0) {
			throw new RuntimeException("List Error: deleteFront() called on empty List or cursor points to null");
		}
		Node temp;
		temp = front.next;
		front = temp;
		front.prev = null;
		temp = null;
		length--;
		if (getIndex() != -1) {
			index--;
		}
	}
	
	// Deletes the back element in this List. Pre: length()>0
	void deleteBack(){  
		if (length()<=0 || getIndex()<0) {
			throw new RuntimeException("List Error: deleteBack() called on empty List or cursor points to null");
		}
		Node temp = null;
		temp = back.prev;
		back = temp;
		back.next = null;
		temp = null;
		length--;
	}
	
	//Deletes cursor element in this List. Cursor is undefined. Pre: length()>0, getIndex()>=0
	void delete() {
		Node be4;
		Node l8r;
		if (length()<=0 || getIndex()<0) {
			throw new RuntimeException("List Error: delete() called on empty List or cursor points to null");
		}
		if(getIndex() == 0){
			deleteFront();
			length++;
		} else if(getIndex() == length()-1){
			deleteBack();
			length++;
		} else {
			be4 = cursor.prev;
			l8r = cursor.next;
			be4.next = l8r;
			l8r.prev = be4;
		}
		cursor = null;
		length--;
	}
	
	// Overides Object's toString method. Creates a string consisting of a space separated sequence of integers with front on the left and back on the right. The cursor is ignored.
	public String toString(){ 
	     String s = "";
	     for(Node N=front; N!=null; N=N.next){
	        //s += N.MatrixEntry + "\n";
	        s += N.MatrixEntry + " ";
	     }
	     return s;
	  }
}
	// Returns a new list representing the same intefer sequence as this list. The cursor in the new list is undefined, regardless of the state of the cursor in this List. This List is unchanged.
	/*List copy() {
		List sb = new List();
		moveTo(0);
		while(getIndex() < length()-1){
			sb.append(getElement());
			moveNext();
		}
		sb.append(getElement());
		
		cursor = null;
		sb.cursor = null;
		return sb;
	} */




